app/forms/provider_interface/offer_wizard.rb

```diff
@@ -38,14 +45,55 @@ def valid_for_current_step?
       valid?(current_step.to_sym)
     end
 
-    def next_step
-      index = STEPS[decision.to_sym].index(current_step.to_sym)
+    def next_step(step = current_step)
+      index = STEPS[decision.to_sym].index(step.to_sym)
+      return unless index
+
+      current_step = STEPS[decision.to_sym][index + 1]
+
+      return save_and_go_to_next_step(current_step) if current_step.eql?(:providers) && available_providers.one?
+      return save_and_go_to_next_step(current_step) if current_step.eql?(:courses) && available_courses.one?
+      return save_and_go_to_next_step(current_step) if current_step.eql?(:study_modes) && available_study_modes.one?
+      return save_and_go_to_next_step(current_step) if current_step.eql?(:locations) && available_course_options.one?
+
+      current_step
+    end
 
-      STEPS[decision.to_sym][index + 1] if index
+    def previous_step
+      path_history[path_history.rindex(current_step) - 1]
     end
 
   private
 
+    def save_and_go_to_next_step(step)
+      attrs = { provider_id: available_providers.first.id } if step.eql?(:providers)
+      attrs = { course_id: available_courses.first.id } if step.eql?(:courses)
+      attrs = { study_mode: available_study_modes.first } if step.eql?(:study_modes)
+      attrs = { course_option_id: available_course_options.first.id } if step.eql?(:locations)
+
+      assign_attributes(last_saved_state.deep_merge(attrs))
+      save_state!
+
+      next_step(step)
+    end
+
+    def available_study_modes
```

Comment:
we seem to be defining those values twice, once in the controller and once here, I was wondering if we only defined them once in the controller, and passed them in as attributes to wizard instead as well as the component?

Reply:
The problem with that is that we don't always need them, neither in the controller or in the wizard. That would force us to retrieve all the values on all steps.

Reply:
We looked a little bit into this and it feels very unnatural to do it in the controller. Besides having to load upfront information, we also rely on the wizard to validate the different information we provide e.g.  the wizard needs to be initialised with a different set of parameters, then we need to check for the validity and then, any of these details can be retrieved. We also need to know the selected option in order to retrieve the next set of information, e.g. if a provider has a single course then we need to check for that before we can retrieve the course options. That moves a lot of the logic to the controller making it do a lot of extra things we can't properly test for.